File: src/main/java/rars/venus/FileStatus.java
package rars.venus;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import rars.Globals;

import java.io.File;

/**
 * Used to store and return information on the status of the current ASM file
 * that
 * is being edited in the program.
 *
 * @author Team JSpim
 */
public final class FileStatus {

    public static @Nullable File systemFile = null;
    private static @NotNull FileStatus.State systemState = State.NO_FILE;
    private static boolean systemAssembled = false;
    private @NotNull FileStatus.State status;
    private @Nullable File file;

    /**
     * Create a FileStatus object with FileStatus.NO_FILE for status and null for
     * file getters.
     */
    public FileStatus() {
        this(State.NO_FILE, null);
    }

    /**
     * Create a FileStatus object with given status and file pathname.
     *
     * @param status
     *     Initial file status. See FileStatus static constants.
     * @param file
     *     File object representing the file.
     */
    public FileStatus(final @NotNull FileStatus.State status, final @Nullable File file) {
        this.status = status;
        this.file = file;
    }

    /**
     * Get file status
     *
     * @return file status EDITED, RUNNABLE, etc, see list above
     */
    public static @NotNull FileStatus.State getSystemState() {
        return systemState;
    }

    /**
     * Set file status. Also updates menu state accordingly.
     *
     * @param newStatus
     *     New status: EDITED, RUNNABLE, etc, see list above.
     */
    public static void setSystemState(final @NotNull FileStatus.State newStatus) {
        systemState = newStatus;
        if (Globals.GUI != null) {
            Globals.GUI.setMenuState(systemState);
        }
    }

    /**
     * Tells whether the file has been assembled.
     *
     * @return Boolean value that is true if the ASM file has been assembled.
     */
    public static boolean isAssembled() {
        return systemAssembled;
    }

    /**
     * Changes the value of assenbked to the parameter given.
     *
     * @param b
     *     boolean variable that tells what to set assembled to.
     */
    public static void setAssembled(final boolean b) {
        systemAssembled = b;
    }

    /**
     * Resets all the values in FileStatus
     */
    public static void reset() {
        systemState = State.NO_FILE;
        systemAssembled = false;
        systemFile = null;
    }

    /**
     * Get editing status of this file.
     *
     * @return current editing status. See FileStatus static constants.
     */
    public @NotNull FileStatus.State getFileStatus() {
        return this.status;
    }

    /**
     * Set editing status of this file. See FileStatus static constants.
     *
     * @param newStatus
     *     the new status
     */
    public void setFileStatus(final @NotNull FileStatus.State newStatus) {
        this.status = newStatus;
    }

    /**
     * Determine if file is "new", which means created using New but not yet saved.
     * If created using Open, it is not new.
     *
     * @return true if file was created using New and has not yet been saved, false
     * otherwise.
     */
    public boolean isNew() {
        return switch (status) {
            case NEW_NOT_EDITED, NEW_EDITED -> true;
            default -> false;
        };
    }

    /**
     * Determine if file has been modified since last save or, if not yet saved,
     * since
     * being created using New or Open.
     *
     * @return true if file has been modified since save or creation, false
     * otherwise.
     */
    public boolean hasUnsavedEdits() {
        return switch (status) {
            case NEW_EDITED, EDITED -> true;
            default -> false;
        };
    }

    /**
     * Get file name with no path information. See java.io.File.getName()
     *
     * @return file as a String
     */
    public @Nullable File getFile() {
        return this.file;
    }

    public void setFile(final @Nullable File file) {
        this.file = file;
    }

    /**
     * Update static FileStatus fields with values from this FileStatus object
     * To support legacy code that depends on the static.
     */
    public void updateStaticFileStatus() {
        systemState = this.status;
        systemAssembled = false;
        systemFile = this.file;
    }

    public enum State {
        /// Initial state or after close.
        NO_FILE,
        /// New edit window with no edits.
        NEW_NOT_EDITED,
        /// New edit window with unsaved edits.
        NEW_EDITED,
        /// Open/saved edit window with no edits.
        NOT_EDITED,
        /// Open/saved edit window with unsaved edits.
        EDITED,
        /// Successful assembly.
        RUNNABLE,
        /// Execution is under way
        RUNNING,
        /// Execution terminated.
        TERMINATED,
        /// File is being opened.
        OPENING;

        public boolean isAssembled() {
            return switch (this) {
                case RUNNABLE, RUNNING, TERMINATED -> true;
                default -> false;
            };
        }
    }

}

File: src/main/java/rars/venus/FileStatus.kt
package rars.venus

import java.io.File

sealed interface FileStatus {
    sealed interface New : FileStatus {
        val tmpName: String

        data class NotEdited(override val tmpName: String) : New
        data class Edited(override val tmpName: String) : New
    }

    sealed interface Existing : FileStatus {
        val file: File

        data class NotEdited(override val file: File) : Existing
        data class Edited(override val file: File) : Existing
        data class Runnable(override val file: File) : Existing
        data class Running(override val file: File) : Existing
        data class Terminated(override val file: File) : Existing
    }
}

fun FileStatus.copy() = when (this) {
    is FileStatus.New.NotEdited -> this.copy()
    is FileStatus.New.Edited -> this.copy()
    is FileStatus.Existing.NotEdited -> this.copy()
    is FileStatus.Existing.Edited -> this.copy()
    is FileStatus.Existing.Runnable -> this.copy()
    is FileStatus.Existing.Running -> this.copy()
    is FileStatus.Existing.Terminated -> this.copy()
}

fun FileStatus.isNew() = this is FileStatus.New
fun FileStatus.isEdited() = when (this) {
    is FileStatus.New.Edited,
    is FileStatus.Existing.Edited,
        -> true
    else -> false
}

fun FileStatus.isAssembled() = when (this) {
    is FileStatus.Existing.Runnable,
    is FileStatus.Existing.Running,
        -> true
    else -> false
}

val FileStatus.nameForEditorTab: String
    get() = when (this) {
        is FileStatus.New -> tmpName
        is FileStatus.Existing -> file.name
    }

val FileStatus.nameForTitleBar: String
    get() = when (this) {
        is FileStatus.New -> tmpName
        is FileStatus.Existing -> file.absolutePath
    }

fun FileStatus.New.toNotEdited() = FileStatus.New.NotEdited(tmpName)
fun FileStatus.New.toEdited() = FileStatus.New.Edited(tmpName)

fun FileStatus.Existing.toNotEdited() = FileStatus.Existing.NotEdited(file)
fun FileStatus.Existing.toEdited() = FileStatus.Existing.Edited(file)
fun FileStatus.Existing.toRunnable() = FileStatus.Existing.Runnable(file)
fun FileStatus.Existing.toRunning() = FileStatus.Existing.Running(file)
fun FileStatus.Existing.toTerminated() = FileStatus.Existing.Terminated(file)

/*
 * Valid states for a single edit panel:
 * 1. New file, not edited - after creating a new file, but before making
 *    any changes to it.
 * 2. New file, edited - new file, but after making changes to it.
 * 3. Not edited - after opening a file, but before making any changes to it.
 * 4. Edited - after making changes to a file.
 * 
 * What happens when a new file button is clicked?
 * We don't have an actual file associated with such a tab...
 */
